accessor/range.hpp

```diff
@@ -42,6 +42,29 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 namespace gko {
 namespace acc {
+namespace {
+
+
+/**
+ * the default check_if_valid gives true.
+ *
+ * @tparam Ref  the reference type
+ * @tparam Args  the input type
+ */
+template <typename Ref, typename... Args>
+struct check_if_valid : public std::true_type {};
+
+/**
+ * check_if_valid gives false if the decay type of input is the same type as
+ * Ref.
+ *
+ * @tparam Ref  the reference type
+ */
+template <typename Ref>
+struct check_if_valid<Ref, Ref> : public std::false_type {};
```

Comment:
Please move this to the private section of `range<Accessor>`. Currently, it pollutes the namespace and is not descriptive enough to know what it does.
Also, I just realized that `check_if_equal` might be a better name. Anyway, it must either be moved to the `range` class or into a `detail` namespace (with strong preference to the `range` class).

Reply:
should I also change the ginkgo range? or changing this for accessor should be enough

Reply:
If we don't have a problem with the public `range`, I would leave it as it is for now. Should we encounter the same problem there (I assume we would as soon as we use it the same way we use `acc::range`), we can simply copy it from `acc::range`.

Reply:
I see